packages/flutter_driver/lib/src/common/wait.dart

```diff
@@ -0,0 +1,178 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:flutter/scheduler.dart';
+
+/// Base class for a condition that can be waited upon.
+abstract class WaitCondition {
+  /// Gets the current status of the [condition], executed in the context of the
+  /// Flutter app:
+  ///
+  /// - True, if the condition is satisfied.
+  /// - False otherwise.
+  ///
+  /// The future returned by [wait] will complete when this [condition] is
+  /// fulfilled.
+  bool get condition;
+
+  /// Returns a future that completes when [condition] turns true.
+  Future<void> wait();
+
+  /// Serializes the object to JSON.
+  Map<String, dynamic> serialize();
+}
+
+/// Thrown to indicate a JSON serialization error.
+class SerializationException implements Exception {
+  /// Creates a [SerializationException] with an optional error message.
+  SerializationException([this.message]);
+
+  /// The error message, possibly null.
+  final String message;
+
+  @override
+  String toString() => 'SerializationException($message)';
+}
+
+/// A condition that waits until no transient callbacks are scheduled.
+class NoTransientCallbacksCondition implements WaitCondition {
+  /// Creates a [NoTransientCallbacksCondition] instance.
+  const NoTransientCallbacksCondition();
+
+  /// Factory constructor to parse a [NoTransientCallbacksCondition] instance
+  /// from the given JSON map.
+  factory NoTransientCallbacksCondition.deserialize(Map<String, dynamic> json) {
+    if (json['conditionName'] != 'NoTransientCallbacksCondition')
+      throw SerializationException('Error occurred during deserializing the NoTransientCallbacksCondition JSON string: $json');
+    return const NoTransientCallbacksCondition();
+  }
+
+  @override
+  bool get condition => SchedulerBinding.instance.transientCallbackCount == 0;
+
+  @override
+  Future<void> wait() async {
```

Comment:
Not really. If you look at the newly added `FirstFrameRasterizedCondition`, they are quite different.

Reply:
I'd still argue that the default is to wait until the next frame if the condition is currently unmet. FirstFrameRasterizedCondition can still override that to provide a custom impl

Reply:
Discussed offline, and kept it as is. Since the `condition` itself is unknown in the base class, we do not have the knowledge of whether the condition and the new frame rendering is relevant. They are relevant for noTransientCallbacks, or noPendingFrame, but they are not for other conditions like checks for pending platform channel messages, Timers, etc.